### PR TITLE
added Jabber

### PR DESCRIPTION
**Pull request by nassimabedi**

### CODE REVIEW

**Comment by glaslos** [comment]
@nassimabedi can you investigate why the tests are failing?

**Comment by nassimabedi** [comment]
@glaslos Yes, let me check it

**Comment by nassimabedi** [comment]
@glaslos Tests passed.

**Comment by nassimabedi** [comment]
@glaslos : Code clean up.

File: jabber.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	"fmt"
+	"net"
+	"context"
+	"encoding/xml"
+	"bufio"
+	"reflect"
+	"unsafe"
+)
+
+type ServersJabber struct {
```
Comment (glaslos) [inline jabber.go]: Just call this `Server`. The same goes for the `Server` and `Client` struct.

File: jabber.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    XMLName xml.Name `xml:"servers"`
+    Version string   `xml:"version,attr"`
+    Svs     []serverJabber `xml:"server"`
+}
+
+type serverJabber struct {
```
Comment (glaslos) [inline jabber.go]: Export this struct

File: jabber.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    ServerName string `xml:"serverName"`
+    ServerIP   string `xml:"serverIP"`
+}
+
+type JabberClient struct {
+		STo         string   `xml:"to,attr"`
```
Comment (glaslos) [inline jabber.go]: formatting seems off

File: jabber.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	    Version     string   `xml:"version,attr"`
+	    Xmlns       string   `xml:"xmlns,attr"`
```
Comment (glaslos) [inline jabber.go]: Stick to either `Xml` or `XML`

File: jabber.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	    Id          string   `xml:"id,attr"`
+	    XmlnsStream string   `xml:"xmlns stream,attr"`
+	    XMLName     xml.Name `xml:"http://etherx.jabber.org/streams stream"`
+	}
+
+
+func bytesToString(b []byte) string {
```
Comment (glaslos) [inline jabber.go]: This should work better `s := string(byteArray[:n])`

File: jabber.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    bh := (*reflect.SliceHeader)(unsafe.Pointer(&b))
+    sh := reflect.StringHeader{bh.Data, bh.Len}
+    return *(*string)(unsafe.Pointer(&sh))
+}
+
+// parse Jabber client
+func parseJabberClient( dataClient string , g *Glutton) string  {
```
Comment (glaslos) [inline jabber.go]: Please use `go fmt` on your code before committing.

File: jabber.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	
+	v := JabberClient{STo: "none", Version: "none"}
+
+	err := xml.Unmarshal([]byte(dataClient), &v)
+	if err != nil {
+		g.logger.Error(fmt.Sprintf("[Jabber    ] err: %v", err))
```
Comment (glaslos) [inline jabber.go]: You should return here on error.

File: jabber.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+         line []byte
+      )
+
+	r := bufio.NewReader(conn)
+	for i := 1; true; i++ {
+		line, isPrefix, err = r.ReadLine()
+		if err != nil {
+			g.logger.Error(fmt.Sprintf("[Jabber     ] error: %v", err))
+			break
+			return
```
Comment (glaslos) [inline jabber.go]: `break` and `return`?

File: jabber.go (hunk 7 above)
Comment (glaslos) [inline jabber.go]: Please adjust the logging format to match the other handlers.

File: jabber.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+	err := xml.Unmarshal([]byte(dataClient), &v)
+	if err != nil {
+		g.logger.Error(fmt.Sprintf("[Jabber    ] err: %v", err))
+	}
+	
+	g.logger.Info(fmt.Sprintf("[Jabber     ] Jabber STo: %v", v.STo))
+	g.logger.Info(fmt.Sprintf("[Jabber     ] Jabber Version: %v", v.Version))
+	g.logger.Info(fmt.Sprintf("[Jabber     ] Jabber Xmlns: %v", v.Xmlns))
+	g.logger.Info(fmt.Sprintf("[Jabber     ] Jabber XMLName: %v", v.XMLName))
```
Comment (glaslos) [inline jabber.go]: Try to log all the information in one line.

File: jabber.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+	for i := 1; true; i++ {
+		line, isPrefix, err = r.ReadLine()
+		if err != nil {
+			g.logger.Error(fmt.Sprintf("[Jabber     ] error: %v", err))
+			break
+			return
+		}
+		
+		g.logger.Info(fmt.Sprintf("[Jabber     ] Jabber isPrefix: %v", isPrefix))
+		g.logger.Info(fmt.Sprintf("[Jabber     ] Jabber line: %v", line))
```
Comment (glaslos) [inline jabber.go]: You can remove those lines or make the `debug` level.

File: jabber.go [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+func (g *Glutton) HandleJabber(ctx context.Context, conn net.Conn) (err error) {
+	defer func() {
+		err = conn.Close()
+		if err != nil {
+			g.logger.Error(fmt.Sprintf("[Jabber     ]  error: %v", err))
+		}
+	}()
+	
+
+	v := &ServersJabber{Version: "1"}
```
Comment (glaslos) [inline jabber.go]: formatting.

File: jabber.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    v.Svs = append(v.Svs, serverJabber{"Test_VPN", "127.0.0.1"})
+ 
+    output, err := xml.MarshalIndent(v, "  ", "    ")
+    if err != nil {
+        g.logger.Error(fmt.Sprintf("[Jabber     ]  error: %v", err))
+    }
+ 
+    conn.Write(output)
```
Comment (glaslos) [inline jabber.go]: I think this can return an error that should be handled.

File: jabber.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+		}
+	}()
+	
+
+	v := &ServersJabber{Version: "1"}
+    v.Svs = append(v.Svs, serverJabber{"Test_VPN", "127.0.0.1"})
+ 
+    output, err := xml.MarshalIndent(v, "  ", "    ")
+    if err != nil {
+        g.logger.Error(fmt.Sprintf("[Jabber     ]  error: %v", err))
```
Comment (glaslos) [inline jabber.go]: return on error here

File: jabber.go [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    }
+ 
+    conn.Write(output)
+
+    msg, err := readMsgJabber(conn, g)
+    if err != nil {
+		g.logger.Error(fmt.Sprintf("[Jaber     ] error: %v", err))
```
Comment (glaslos) [inline jabber.go]: return on error here

**Comment by nassimabedi** [comment]
@glaslos : Thank you SO much :) I 'll fix all of the comments.

File: jabber.go (hunk 1 above)
Comment (nassimabedi) [inline jabber.go]: @glaslos Server xml's format is different due to client side portion at jabber service.

server code:
`<?xml version="1.0" encoding="UTF-8"?>
  <servers version="1">
      <server>
          <serverName>Shanghai_VPN</serverName>
          <serverIP>127.0.0.1</serverIP>
      </server>
      <server>
          <serverName>Beijing_VPN</serverName>
          <serverIP>127.0.0.2</serverIP>
      </server>
  </servers>`
client code :
`<?xml version='1.0' ?>
<stream:stream to='192.168.58.137' xmlns='jabber:client'  xmlns:stream='http://etherx.jabber.org/streams' version='1.0'></stream:stream>`

File: jabber.go (hunk 3 above)
Comment (nassimabedi) [inline jabber.go]: done

File: jabber.go (hunk 4 above)
Comment (nassimabedi) [inline jabber.go]: done

File: jabber.go (hunk 5 above)
Comment (nassimabedi) [inline jabber.go]: done

File: jabber.go (hunk 6 above)
Comment (nassimabedi) [inline jabber.go]: done

File: jabber.go (hunk 7 above)
Comment (nassimabedi) [inline jabber.go]: done

Comment (nassimabedi) [inline jabber.go]: done

File: jabber.go (hunk 9 above)
Comment (nassimabedi) [inline jabber.go]: done

File: jabber.go (hunk 8 above)
Comment (nassimabedi) [inline jabber.go]: done

File: jabber.go (hunk 10 above)
Comment (nassimabedi) [inline jabber.go]: done

File: jabber.go (hunk 11 above)
Comment (nassimabedi) [inline jabber.go]: done

File: jabber.go (hunk 13 above)
Comment (nassimabedi) [inline jabber.go]: done

File: jabber.go (hunk 12 above)
Comment (nassimabedi) [inline jabber.go]: done

File: jabber.go (hunk 14 above)
Comment (nassimabedi) [inline jabber.go]: done.

**Comment by glaslos** [comment]
Can you use go fmt to format the code and also rebase your PR? I really want to get this merged some day :)

**Comment by nassimabedi** [comment]
Sure 😄

**Comment by glaslos** [comment]
Closing this one due to the new PR

**Pull request closed without merging**